ducktape/cluster/ansible.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2022 Redpanda Data Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Any, Optional
+from ducktape.cluster.json import JsonCluster
+from ansible.inventory.manager import InventoryManager
+from ansible.parsing.dataloader import DataLoader
+from ansible.vars.manager import VariableManager
+
+import os
+import json
+
+
+class AnsibleCluster(JsonCluster):
+    """
+    An implementation of Cluster that uses a set of VMs listed in an ansible
+    inventory file. Hosts will be taken from the inventory group name
+    `node_group`. See tests/cluster/check_ansible.py for example inventory file.
+    """
+
+    def __init__(self, inv_filename: str, cluster_file="cluster.json",
+                 node_group="kafka"):
+
+        self.ansible_file = inv_filename
+        self.cluster_file = cluster_file
+        self.node_group = node_group
+
+        self.log = logging.getLogger(__file__)
+        self.log.setLevel(logging.DEBUG)
+
+        # translate to known json format
+        loader = DataLoader()
+        self.inventory = InventoryManager(loader=loader, sources=inv_filename)
```

Comment:
Do you know if this supports both .ini and YAML inventory formats? I added YAML output to the terraform scripts since it was easier for me to consume it as I didn't want to pull in all the ansible deps (but maybe I should have), so I'm wondering if standardize on YAML if this manager supports it (I think it does, via autodetection?).

Reply:
Similar question. The inventory file is something we generate through terraform / or generally control the creation of. For example, we could simultaneously generate a INI formatted file for Ansible and a JSON formatted file for which there is already a load available in ducktape. However, it also seems like Ansible might be able to read a YAML inventory file out-of-the-box?

The top-level idea here is: it would be nice to upstream these changes into ducktape so we don't have to maintain them. And I think selling a dependency on all of the python ansible library will be harder.

Reply:
IIRC, this uses the same code as ansible to load inventory, so it supports the same formats.

I also considered the two other options (1) just have terraform spit out a cluster.json via a template (2) write a separate translator tool. Agreed we don't really want to maintain a fork of ducktape--but we need to agree on a policy and stick with it. (i.e. "all changes must flow through upstream first or have a special string in the commit message, etc.). Agree the dependency on ansible is unfortunate--I had the exact same thoughts.
I *did* learn some basic ducktape hacking skills and code familiarity--which I think was the real value here. I'm not attached to this.. After doing it I think we might want to just spit out a cluster.json from terraform. I started on that--but did not finish--so I'm not sure if there are any issues there.

Reply:
> (i.e. "all changes must flow through upstream first or have a special string in the commit message, etc.)

IME it can be too much of a hassle to always push patches through upstream in order to use them. But authoring the PR as if it is going upstream is useful because we can merge locally for immediate use and submit upstream to deal with that interaction asynchronously.

> (2) write a separate translator tool

this seems like the cleanest option because it can be invoked in whatever tools we build. we probably won't ever be invoking ducktape manually its such a big CLI invocation.